core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/PreAggregateCaseAggregations.java

```diff
@@ -363,6 +367,14 @@ private Optional<CaseAggregation> extractCaseAggregation(Symbol aggregationSymbo
             return Optional.empty();
         }
 
+        Expression whenOperand = caseExpression.getWhenClauses().getFirst().getOperand();
+        Expression whenResult = caseExpression.getWhenClauses().getFirst().getResult();
+        if (wrappedByCast ||
+                !(caseExpression.getWhenClauses().getFirst().getResult() instanceof SymbolReference ||
+                        caseExpression.getWhenClauses().getFirst().getResult() instanceof Literal)) {
+            // Wrap the result expression with IF to retain CASE's conditional property during pre-aggregation
+            whenResult = new IfExpression(whenOperand, whenResult, null);
```

Comment:
This is probably not sufficient, e.g: query
```
select
max(case(when col = 1 then a / b end) i,
max(case(when col = 2 then a / b end) j
```
will be rewritten to something like
```
select
  max(case(when col = 1 then pre_aggregate end) i,
  max(case(when col = 2 then pre_aggregate end) i,
from (
  select max(case col = 1 then a / b end) from table group by col
```

Hence, wrapping with
`case col = 1 then a / b end`
is not sufficient because `pre_aggregate` is used in two aggregations.
Proper rewrite should be `case col = 1 OR col = 2 then a / b end`.

Could you test or modify existing tests to test that case?

Reply:
No it's covered. The rewrite will project two independent expressions, which are used for the subsequent preaggregations respectively.
```
expr_6 := (CASE WHEN (col = 1) THEN (a / b) END)
expr_8 := (CASE WHEN (col = 2) THEN (a / b) END)
```
Please see this test case https://github.com/trinodb/trino/pull/20652/commits/77c8ef22dff713d89a97e5a2e941274f17a6826a#diff-ea0b3e8d29bf709db8390ecc5e4a26cf4367a66251f57eb072e1cf424a0c887bR66

Reply:
OK. This means there might be a regression for such case expressions because now there will be two pre-aggregations instead of one, so the gains from pre-aggregations are essentially gone.

If you look at orginal PR, then there were planner tests (which are now removed, cc @martint) and tpc-ds q66 is definitely affected by this change (I expect quite significant regression there).

I think we should use something like ```case col = 1 OR col = 2 then a / b end``` to avoid creating extra pre-aggregates.

Also, could you benchmark q66 before/after change?

Reply:
Your concern is valid. I updated the patch to merge the eligible conditions so that we can still get a small subset of aggregation expressions. For Q66, the preaggreation part of the plan now looks like
```
expr_222 := (CASE WHEN ("d_moy" IN (1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12)) THEN ("ws_net_paid" * CAST("ws_quantity" AS decimal(10, 0))) END)                                               >
expr_220 := (CASE WHEN ("d_moy" IN (1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12)) THEN ("ws_ext_sales_price" * CAST("ws_quantity" AS decimal(10, 0))) END)
```

Reply:
Could you apply this fix on top of https://github.com/trinodb/trino/tree/345652efced5f6c9b3491adff91156b67df167fb and run tests from https://github.com/trinodb/trino/tree/345652efced5f6c9b3491adff91156b67df167fb/testing/trino-benchto-benchmarks/src/test/java/io/trino/sql/planner to make sure other plans did not rotate?

Reply:
I applied the patch to version 391 and ran these 4 tests. They all passed.
* TestHivePartitionedTpcdsCostBasedPlan.java
* TestHivePartitionedTpchCostBasedPlan.java
* TestHiveTpcdsCostBasedPlan.java
* TestHiveTpchCostBasedPlan.java